File: Play-with-Algorithms/Play-with-Algorithms-Java/src/main/java/com/matrix/algorithms/SelectionSort.java
package com.matrix.algorithms;


/**
 * 选择排序
 */
public class SelectionSort {


    // 算法类不允许产生任何实例
    private SelectionSort() {
    }

    private static void sort(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n; i++) {
            // 寻找[i,n)区间里最小值的索引
            int minIndex = i;
            for (int j = i + 1; j < n; j++) {
                if (arr[j] < arr[minIndex]) {
                    minIndex = j;
                }
            }
            swap(arr, i, minIndex);
        }
    }

    private static void swap(int[] arr, int i, int j) {
        int t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    public static void main(String[] args) {
        int[] arr = {20, 14, 77, 54, 65, 84, 75, 22, 65, 54, 11};
        SelectionSort.sort(arr);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            System.out.print(" ");
        }
    }

}

File: Play-with-Algorithms/Play-with-Algorithms-Python/com.matrix.algorithms/SelectionSort.py
# !/usr/bin/env python
# -*- coding:utf-8 -*-
# @Copyright (C), 2017, matrix


class SelectionSort(object):
    """
    选择排序
    """

    def __init__(self):
        pass

    def sort(self, arr):
        for i in range(0, len(arr)):
            minIndex = i
            for j in range(i + 1, len(arr)):
                if arr[j] < arr[minIndex]:
                    minIndex = j
            self.swap(arr, i, minIndex)

    def swap(self, arr, i, j):
        t = arr[i]
        arr[i] = arr[j]
        arr[j] = t


if __name__ == "__main__":
    arr = [20, 14, 77, 58, 65, 84, 75, 22, 62, 54, 11]
    i = SelectionSort()
    i.sort(arr)
    print(arr)
